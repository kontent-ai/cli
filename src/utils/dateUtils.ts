export const formatDateForFileName = (date: Date) =>
    `${date.getUTCFullYear()}-` +
    `${('0' + (date.getUTCMonth() + 1)).slice(-2)}-` +
    `${('0' + date.getUTCDate()).slice(-2)}-` +
    `${('0' + date.getUTCHours()).slice(-2)}-` +
    `${('0' + date.getUTCMinutes()).slice(-2)}-` +
    `${('0' + date.getUTCSeconds()).slice(-2)}-`;
