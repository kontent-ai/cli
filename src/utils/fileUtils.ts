import fs, { PathLike } from 'fs';
import * as path from 'path';

export const fileExists = (filePath: PathLike): boolean => {
    return fs.existsSync(filePath);
};

export const isAllowedExtension = (filename: string): boolean => {
    const extension = path.extname(filename);
    return ['.js', ''].includes(extension);
};

export const getFileWithExtension = (filename: string, defaultExtension: string = '.js'): string => {
    const hasFileExtension = path.extname(filename);

    if (hasFileExtension) {
        const normalized = filename.split(defaultExtension).slice(0, -1).join(defaultExtension);
        return normalized + defaultExtension;
    } else {
        return filename + defaultExtension;
    }
};

export const getFileBackupName = () => {
    const currentDate = new Date();
    const formatted = `${currentDate.getFullYear()}-${currentDate.getMonth()}-${currentDate.getDate()}-${currentDate.getTime()}`;
    return `backup-${formatted}`;
};
