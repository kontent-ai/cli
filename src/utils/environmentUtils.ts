import chalk from 'chalk';
import fs from 'fs';
import * as path from 'path';
import { IEnvironmentsConfig } from '../models/environment';
import { fileExists } from './fileUtils';

const environmentsConfigName = '.environments.json';

export const saveEnvironmentConfig = (name: string, projectId: string, apiKey: string) => {
    const environments = getEnvironmentsConfig();
    if (environments[name]) {
        console.log(chalk.yellow(`The \"${name}\" environment already exists and will be overwritten.`));
    }

    environments[name] = { projectId, apiKey };
    const environmentsJSON = JSON.stringify(environments, null, 2);

    saveEnvironmentData(environmentsJSON, name);
};

const saveEnvironmentData = (data: string, name: string): void => {
    const configsFilepath = getEnvironmentConfigFilepath();

    fs.writeFile(configsFilepath, data, { flag: 'w' }, (error) => {
        if (error) {
            console.error(chalk.red(error.stack || error.message));
        } else {
            console.log(chalk.green(`The environment ${name} (\"${configsFilepath}\") was updated.`));
        }
    });
};

const getEnvironmentConfigFilepath = (): string => path.join(process.cwd(), environmentsConfigName);

export const getEnvironmentsConfig = (): IEnvironmentsConfig => {
    const environmentConfigFilepath = getEnvironmentConfigFilepath();

    if (!fileExists(environmentConfigFilepath)) {
        return {};
    }

    const environmentsBuffer = fs.readFileSync(environmentConfigFilepath);

    return JSON.parse(environmentsBuffer.toString());
};

export const environmentConfigExists = () => {
    const environmentConfigFilepath = getEnvironmentConfigFilepath();

    return fileExists(environmentConfigFilepath);
};
