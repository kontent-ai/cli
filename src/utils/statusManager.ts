import { IMigrationStatus, IStatus } from '../models/status';
import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { fileExists } from './fileUtils';
import * as path from 'path';
import { loadStatusPlugin } from './status/statusPlugin';

const migrationStatusFilename = 'status.json';
const statusDirectoryName = 'status';
const statusImplementationFilename = 'statusImpl.ts';
let status: IStatus = {};

const updateMigrationStatus = async (projectId: string, migrationStatus: IMigrationStatus) => {
    let projectMigrationsHistory = status[projectId];

    if (projectMigrationsHistory === undefined) {
        projectMigrationsHistory = status[projectId] = [];
    }

    const previousMigrationStatusIndex = projectMigrationsHistory.findIndex((executedMigration) => executedMigration.name === migrationStatus.name);
    if (previousMigrationStatusIndex > -1) {
        projectMigrationsHistory[previousMigrationStatusIndex] = migrationStatus;
    } else {
        projectMigrationsHistory.push(migrationStatus);
    }

    await saveStatusFile();
};

export const markAsCompleted = async (projectId: string, name: string, order: number | Date) => {
    const migrationStatus = {
        name,
        order,
        success: true,
        time: new Date(Date.now()),
    };

    await updateMigrationStatus(projectId, migrationStatus);
};

const saveStatusFile = async () => {
    const statusJSON = JSON.stringify(status, null, 2);

    if (fileExists(getStatusImplementationFilePath())) {
        try {
            const file = await loadStatusPlugin(getStatusImplementationFilePath().slice(0, -3) + '.js');
            await file.saveStatus(statusJSON);
        } catch (e) {
            console.log(`Could not load the plugin due to ${e}. Fallbacking to status.json`);
        } finally {
            return;
        }
    }
    saveStatusToFile(statusJSON);
};

const getMigrationStatus = (migrationName: string, projectId: string): IMigrationStatus | undefined => {
    const projectStatus = status[projectId];
    if (projectStatus == null) {
        return undefined;
    }

    return projectStatus.find((migrationStatus) => migrationStatus.name === migrationName);
};

export const wasSuccessfullyExecuted = (migrationName: string, projectId: string): boolean => {
    const result = getMigrationStatus(migrationName, projectId);
    return !!result && result.success;
};

const getStatusFilepath = (): string => {
    return path.join(process.cwd(), migrationStatusFilename);
};

export const loadMigrationsExecutionStatus = async () => {
    if (fileExists(getStatusImplementationFilePath())) {
        try {
            const file = await loadStatusPlugin(getStatusImplementationFilePath().slice(0, -3) + '.js');
            status = await file.readStatus();
        } catch (e) {
            console.log(`Could not load the plugin due to ${e}`);
        } finally {
            return;
        }
    }
    status = readFromStatus();
};

const readFromStatus = (): IStatus => {
    const statusFilepath = getStatusFilepath();
    if (!fileExists(statusFilepath)) {
        return {};
    }

    try {
        const projectsMigrationStatuses = readFileSync(getStatusFilepath()).toString();

        return JSON.parse(projectsMigrationStatuses);
    } catch (error) {
        console.warn(`Status JSON file is invalid because of ${error instanceof Error ? error.message : 'unknown error.'}. Continuing with empty status.`);
        return {};
    }
};

const saveStatusToFile = (data: string): void => {
    const statusFilepath = getStatusFilepath();

    try {
        writeFileSync(statusFilepath, data, { flag: 'w' });
        console.log(`Status file was updated see ${statusFilepath}`);
    } catch (error) {
        console.error(`Status file save failed, because of ${error instanceof Error ? error.message : 'unknown error.'}`);
    }
};

const getStatusImplementationFilePath = () => path.join(process.cwd(), statusDirectoryName, statusImplementationFilename);

const getStatusImplementationDirectoryPath = () => path.join(process.cwd(), statusDirectoryName);

export const createStatusImplementationFile = () => {
    const statusImplementationPath = getStatusImplementationFilePath();

    if (fileExists(statusImplementationPath)) {
        console.error(`File ${statusImplementationPath} already exists`);
    }

    try {
        if (!fileExists(getStatusImplementationDirectoryPath())) {
            mkdirSync(getStatusImplementationDirectoryPath());
        }

        writeFileSync(statusImplementationPath, statusImplementationTemplate, { flag: 'w' });
        console.log(`Status.ts file was created see ${statusImplementationPath}`);
    } catch (error) {
        console.error(`Status.ts file creation failed, because of ${error instanceof Error ? error.message : 'unknown error.'}`);
    }
};

const statusImplementationTemplate = `
import type {IStatus} from "@kontent-ai/cli";

export const saveStatus = async (data: string) => {

}

export const readStatus = async (): IStatus => {

}
`;
