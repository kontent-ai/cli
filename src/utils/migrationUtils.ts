import { ManagementClient } from '@kontent-ai/management-sdk';
import chalk from 'chalk';
import path from 'path';
import fs from 'fs';
import { listFiles } from './fileUtils';
import { TemplateType } from '../models/templateType';
import { MigrationModule } from '../types';
import { IMigration } from '../models/migration';
import { markAsCompleted, wasSuccessfullyExecuted } from './statusManager';

export const getMigrationDirectory = (): string => {
    const migrationDirectory = 'Migrations';
    return path.join(process.cwd(), migrationDirectory);
};

export const getMigrationFilepath = (filename: string): string => {
    return path.join(getMigrationDirectory(), filename);
};

const ensureMigrationsDirectoryExists = () => {
    const directory = getMigrationDirectory();
    if (!fs.existsSync(directory)) {
        fs.mkdirSync(directory);
    }
};

export const saveMigrationFile = (migrationName: string, migrationData: string, templateType: TemplateType): string => {
    ensureMigrationsDirectoryExists();
    const fileExtension = templateType === TemplateType.TypeScript ? '.ts' : '.js';
    const migrationFilepath = getMigrationFilepath(migrationName + fileExtension);

    try {
        fs.writeFileSync(migrationFilepath, migrationData);
        console.log(chalk.green(`Migration template ${migrationName} (${migrationFilepath}) was generated.`));
    } catch (e) {
        console.error(`Couldn't save the migration.`, e.message);
    }

    return migrationFilepath;
};

export const runMigration = async (migration: IMigration, client: ManagementClient, projectId: string): Promise<number> => {
    console.log(`Running the ${migration.name} migration.`);

    let isSuccess = true;

    try {
        await migration.module.run(client).then(() => {
            markAsCompleted(projectId, migration.name, migration.module.order);
        });
    } catch (e) {
        console.error(chalk.redBright('An error occurred while running migration:'), chalk.yellowBright(migration.name), chalk.redBright('see the output from running the script.'));

        let error = e;
        if (e.originalError !== undefined) {
            console.group('Error details');
            console.error(chalk.redBright('Message:'), e.message);
            console.error(chalk.redBright('Code:'), e.errorCode);
            console.error(chalk.redBright('Validation Errors:'), e.validationErrors);
            console.groupEnd();
            console.log();
            console.group('Response details:');
            console.error(chalk.redBright('Message:'), e.originalError.message);
            console.groupEnd();

            error = e.originalError;
        } else {
            console.group('Error details');
            console.error(chalk.redBright('Message:'), e.message);
            console.groupEnd();
        }

        const requestConfig = error.config;
        if (requestConfig) {
            const bodyData = JSON.parse(requestConfig.data || {});
            console.log();
            console.group('Request details:');
            console.error(chalk.yellow('URL:'), requestConfig.url);
            console.error(chalk.yellow('Method:'), requestConfig.method);
            console.error(chalk.yellow('Body:'), bodyData);
            console.groupEnd();
        }

        isSuccess = false;
    }

    if (!isSuccess) {
        return 1;
    }

    console.log(chalk.green(`The \"${migration.name}\" migration on a project with ID \"${projectId}\" executed successfully.`));
    return 0;
};

export const generateTypedMigration = (): string => {
    return `import {MigrationModule} from "@kontent-ai/cli";

const migration: MigrationModule = {
    order: 1,
    run: async (apiClient) => {
    },
};

export default migration;
`;
};

export const generatePlainMigration = (): string => {
    return `
const migration = {
    order: 1,
    run: async (apiClient) => {
    },
};

module.exports = migration;
`;
};

export const createMigration = (migrationName: string, templateType: TemplateType): string => {
    ensureMigrationsDirectoryExists();
    const generatedMigration = templateType === TemplateType.TypeScript ? generateTypedMigration() : generatePlainMigration();

    return saveMigrationFile(migrationName, generatedMigration, templateType);
};

export const getDuplicates = <T extends any>(array: T[], key: (obj: T) => number): T[] => {
    const allEntries = new Map<number, T[]>();
    let duplicates: T[] = [];

    for (const item of array) {
        const itemKey = key(item);
        const prevItem = allEntries.get(itemKey) || [];
        allEntries.set(itemKey, prevItem.concat(item));
    }

    for (const [, value] of allEntries.entries()) {
        if (value.length > 1) {
            duplicates = duplicates.concat(value);
        }
    }

    return duplicates;
};

export const getMigrationsWithInvalidOrder = <T extends any>(array: T[]): T[] => {
    const migrationsWithInvalidOrder: T[] = [];

    for (const migration of array) {
        if (!Number.isInteger(migration.module.order) || Number(migration.module.order) < 0) {
            migrationsWithInvalidOrder.push(migration);
        }
    }

    return migrationsWithInvalidOrder;
};

export const loadModule = async (migrationFile: string): Promise<MigrationModule> => {
    const migrationPath = getMigrationFilepath(migrationFile);

    return await import(migrationPath)
        .then(async (module) => {
            const importedModule: MigrationModule = module.default;
            return importedModule;
        })
        .catch((error) => {
            throw new Error(chalk.red(`Couldn't import the migration script from \"${migrationPath}"\ due to an error: \"${error.message}\".`));
        });
};

export const loadMigrationFiles = async (): Promise<IMigration[]> => {
    const migrations: IMigration[] = [];

    const files = listFiles('.js');

    for (const file of files) {
        migrations.push({ name: file.name, module: await loadModule(file.name) });
    }

    return migrations.filter(String);
};

export const getSuccessfullyExecutedMigrations = (migrations: IMigration[], projectId: string): IMigration[] => {
    const alreadyExecutedMigrations: IMigration[] = [];

    // filter by execution status
    for (const migration of migrations) {
        if (wasSuccessfullyExecuted(migration.name, projectId)) {
            alreadyExecutedMigrations.push(migration);
        }
    }

    return alreadyExecutedMigrations.filter(String);
};
