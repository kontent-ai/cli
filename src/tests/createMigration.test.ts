import { generatePlainMigration, generateTypedMigration } from '../utils/migrationUtils';

describe('Migration generation', () => {
    it('Creates new typescript migration with name', () => {
        const migrationContent = generateTypedMigration();
        expect(migrationContent).toMatchSnapshot();
    });

    it('Creates new javascript migration with name', () => {
        const migrationContent = generatePlainMigration();

        expect(migrationContent).toMatchSnapshot();
    });
});
