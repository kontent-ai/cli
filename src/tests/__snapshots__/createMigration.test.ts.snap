// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migration generation Creates new javascript migration with name 1`] = `
"
const migration = {
    order: 1,
    run: async (apiClient) => {
    },
};

module.exports = migration;
"
`;

exports[`Migration generation Creates new typescript migration with name 1`] = `
"import {MigrationModule} from \\"@kontent-ai/cli\\";

const migration: MigrationModule = {
    order: 1,
    run: async (apiClient) => {
    },
};

export default migration;
"
`;
